### PR TITLE
Do not assume storageclass is still in-tree after csi migration

### LINKED ISSUE #94189: Unable to resize PVC after CSI migration
**What happened**:
I tried to resize PVC which is migrated to CSI but the operation does not succeed.

**What you expected to happen**:
I would expect to be able to resize PVC after migration.

**How to reproduce it (as minimally and precisely as possible)**:

1. Setup a kubernetes v1.17.11 cluster with in-tree volume plugins

2. Create PVC

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
allowVolumeExpansion: true
metadata:
  name: default
parameters:
  type: gp2
provisioner: kubernetes.io/aws-ebs
reclaimPolicy: Delete
volumeBindingMode: Immediate
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: ebs-claim
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 4Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: app
spec:
  containers:
  - name: app
    image: centos
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: persistent-storage
      mountPath: /data
  volumes:
  - name: persistent-storage
    persistentVolumeClaim:
      claimName: ebs-claim
```

3. Ensure that the PVC is successfully `Bound` with the in-tree provisioner

```
$ k describe pvc
Name:          ebs-claim
Namespace:     default
StorageClass:  default
Status:        Bound
Volume:        pvc-815ca829-865b-459f-bacc-139d3266128c
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/aws-ebs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      4Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    app
Events:
  Type    Reason                 Age   From                         Message
  ----    ------                 ----  ----                         -------
  Normal  ProvisioningSucceeded  41s   persistentvolume-controller  Successfully provisioned volume pvc-815ca829-865b-459f-bacc-139d3266128c using kubernetes.io/aws-ebs
```

4. Upgrade to v1.18.8, install aws-ebs-csi-driver and enable the CSI migration

More details on how we perform the CSI migration can be found in this document - https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/csimigration/README.md.

5. Ensure that the PVC is migrated (`pv.kubernetes.io/migrated-to` annotation is set)

```
$ k describe pvc
Name:          ebs-claim
Namespace:     default
StorageClass:  default
Status:        Bound
Volume:        pvc-815ca829-865b-459f-bacc-139d3266128c
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               pv.kubernetes.io/migrated-to: ebs.csi.aws.com
               volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/aws-ebs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      4Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    app
Events:
  Type    Reason                 Age    From                         Message
  ----    ------                 ----   ----                         -------
  Normal  ProvisioningSucceeded  8m22s  persistentvolume-controller  Successfully provisioned volume pvc-815ca829-865b-459f-bacc-139d3266128c using kubernetes.io/aws-ebs
```

6. Edit the PVC and try to resize it

7. Ensure that the resize fails with:

```
$ k describe pvc
Name:          ebs-claim
Namespace:     default
StorageClass:  default
Status:        Bound
Volume:        pvc-815ca829-865b-459f-bacc-139d3266128c
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               pv.kubernetes.io/migrated-to: ebs.csi.aws.com
               volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/aws-ebs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      4Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    app
Events:
  Type     Reason                 Age               From                         Message
  ----     ------                 ----              ----                         -------
  Normal   ProvisioningSucceeded  9m10s             persistentvolume-controller  Successfully provisioned volume pvc-815ca829-865b-459f-bacc-139d3266128c using kubernetes.io/aws-ebs
  Normal   ExternalExpanding      2s (x11 over 7s)  volume_expand                CSI migration enabled for ebs.csi.aws.com; waiting for external resizer to expand the pvc
  Warning  ExternalExpanding      2s (x11 over 7s)  volume_expand                error getting CSI driver name for pvc default/ebs-claim, with error could not find CSI Driver name for plugin ebs.csi.aws.com
```

Ensure that the annotation `volume.kubernetes.io/storage-resizer` is never set to the PVC.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.17.11 + v1.18.8
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools: https://github.com/gardener/gardener + https://github.com/gardener/gardener-extension-provider-aws
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by ialidzhikov**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Volume expand controller is responsible for setting the `volume.kubernetes.io/storage-resizer` annotation for migrated volume. However currently the controller relies that the PVC StorageClass will be still the in-tree (legacy) one. Basically it currently tries to fetch the PVC StorageClass, it gets the StorageClass `provisioner` (and assumes that it is the in-tree one), then tries to get the csi plugin name by in-tree plugin name.
Often after csi migration users replace their legacy StorageClasses with the new out-of-tree. If that is the case, the volume_expand controller fails to set the `volume.kubernetes.io/storage-resizer` annotation and volume resize for a migrated volume fails. I couldn't find a statement that deleting the legacy storageclass and recreating the out-of-tree one with the same name is forbidden and not supposed to happen by the csi contract. Feel free to correct me if it is.

With the PR, volume expand controller is adapted to get the in-tree plugin name from the PV spec (not from the PVC StorageClass provisioner).

**Which issue(s) this PR fixes**:

Fixes #94189

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
An issues preventing volume expand controller to annotate the PVC with `volume.kubernetes.io/storage-resizer` when the PVC StorageClass is already updated to the out-of-tree provisioner is now fixed.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by msau42** [comment]
I think there's a general problem that users can recreate storageclasses with the same name, but completely change the contents of the StorageClass, which may invalidate the PV.spec.storageclassname on existing PVs.

This is why we do not rely on StorageClass after PV creation, and perhaps this `StorageClass.allowVolumeExpansion` field is flawed in that regard.

**Comment by ialidzhikov** [comment]
> This is why we do not rely on StorageClass after PV creation,

Yep, I agree. This PR should improve the volume expand controller wrt that.

**Comment by ialidzhikov** [comment]
@gnufied , @leakingtapan , @msau42 did you managed to check this PR?
I want also to cherry-pick this change into v1.17, v1.18, v1.19.

File: pkg/controller/volume/expand/expand_controller.go [hunk 1]
```diff
@@ -290,6 +283,18 @@ func (expc *expandController) syncHandler(key string) error {
...
+		return nil
+	}
+	class, err := expc.classLister.Get(claimClass)
+	if err != nil {
+		klog.V(4).Infof("failed to expand PVC: %s with error: %v", util.ClaimToClaimKey(pvc), err)
+		return nil
+	}
+
+	volumeResizerName := class.Provisioner
 	return expc.expand(pvc, pv, volumeResizerName)
```
Comment (gnufied) [inline pkg/controller/volume/expand/expand_controller.go:298]: spoke with @msau42 offline and I think we don't need to read storageclass at all in expand_controller. Can you just get resizer name from `volumePlugin.GetPluginName()` which should be same as provisioner name for all intree plugin types and for out of tree plugins, we don't really set those here anyways.

Comment (ialidzhikov) [reply pkg/controller/volume/expand/expand_controller.go:277]: Nice suggestion indeed, thank you! I now adapted to `volumePlugin.GetPluginName()` and removed the StorageClass lister. You can have another look.

**Comment by ialidzhikov** [comment]
Can someone please assist with `/ok-to-test`?

**Comment by ialidzhikov** [comment]
@gnufied , did you managed to re-check?

**Comment by ialidzhikov** [comment]
@msau42 , if you don't have objections, can we add `/milestone v1.20`?

**Comment by msau42** [comment]
/milestone v1.20 

Will let @gnufied review

**Comment by ialidzhikov** [comment]
ping @gnufied @deads2k

File: pkg/controller/volume/expand/expand_controller_test.go [hunk 2]
```diff
@@ -59,27 +57,14 @@ func TestSyncHandler(t *testing.T) {
...
-			pv:     getFakePersistentVolume("vol-1", csitranslationplugins.AWSEBSInTreePluginName, "no-sc-pvc-vol-1"),
-			pvc:    getFakePersistentVolumeClaim("no-sc-pvc", "vol-1", "", "no-sc-pvc-vol-1"),
-			pvcKey: "default/no-sc-pvc",
-		},
-		{
-			name:   "when pvc storageclass is missing",
-			pv:     getFakePersistentVolume("vol-2", csitranslationplugins.AWSEBSInTreePluginName, "missing-sc-pvc-vol-2"),
-			pvc:    getFakePersistentVolumeClaim("missing-sc-pvc", "vol-2", "resizable", "missing-sc-pvc-vol-2"),
-			pvcKey: "default/missing-sc-pvc",
-		},
```
Comment (gnufied) [inline pkg/controller/volume/expand/expand_controller_test.go:77]: rather than deleting this test, can we make sure that - this passes now?

**Comment by ialidzhikov** [comment]
@msau42 , I think we are back to you (or someone else that can assist with review/approval).

**Comment by msau42** [comment]
/assign @deads2k 

for controller manager aproval

**Comment by deads2k** [comment]
kcm change lgtm
/approve

**Pull request merged**